core/virtualhostname/virtualhostname.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package virtualhostname
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v5"
+)
+
+const (
+	// ErrNoMatch indicates that the hostname could not be parsed.
+	ErrNoMatch = errors.ConstError("could not parse hostname")
+)
+
+// HostnameTarget defines what kind of infrastructure the user is targeting.
+type HostnameTarget int
+
+const (
+	// MachineTarget defines a machine as the target.
+	MachineTarget HostnameTarget = iota
+
+	// UnitTarget defines a unit (K8s or machine)
+	// as the target.
+	UnitTarget
+
+	// ContainerTarget defines a container within a
+	// unit as the target.
+	ContainerTarget
+)
+
+// Info returns a breakdown of a virtual
+// hostname into its constituent parts.
+// The target field indicates what kind of
+// hostname was parsed which will indicate
+// that some fields are empty.
+type Info struct {
+	target    HostnameTarget
+	modelUUID string
+	machine   int
+	unit      int
+	app       string
```

Comment:
We are only ever interested in surfacing the unit name, ie "mysql/666" etc. So there's no need to store app and unit in their separated out form. Why not just create the unit name string when the reg exp is parsed. Avoids the unnecessary Sprintf each time someone wants to get the unit name etc.

Reply:
Good point, will change.